network/p2p/p2pbuilder/libp2pNodeBuilder.go

```diff
@@ -502,8 +484,21 @@ func DefaultNodeBuilder(
 	builder.SetGossipSubScoreTracerInterval(gossipCfg.ScoreTracerInterval)
 
 	if role != "ghost" {
-		r, _ := flow.ParseRole(role)
+		r, err := flow.ParseRole(role)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse role: %w", err)
+		}
 		builder.SetSubscriptionFilter(subscription.NewRoleBasedFilter(r, idProvider))
+
+		if r == flow.RoleAccess || r == flow.RoleExecution {
+			// Only access and execution nodes need to run DHT;
+			// Access nodes and execution nodes need DHT to run a blob service.
+			// Moreover, access nodes run a DHT to let un-staked (public) access nodes find each other on the public network.
+			builder.SetRoutingSystem(
+				func(ctx context.Context, host host.Host) (routing.Routing, error) {
+					return dht.NewDHT(ctx, host, protocols.FlowDHTProtocolID(sporkId), logger, metricsCfg.Metrics, dht.AsServer())
+				})
+		}
```

Comment:
should this go in scaffold? I feel like node specific behavior doesn't belong in the `DefaultNodeBuilder` method.

Reply:
https://github.com/onflow/flow-go/pull/4800/commits/d0086b9d73ddf688a921e6c0c226404ca6aee41f